op-node/rollup/derive/calldata_source.go

```diff
@@ -15,21 +14,60 @@ import (
 // CalldataSource readers raw transactions from a given block & then filters for
 // batch submitter transactions.
 // This is not a stage in the pipeline, but a wrapper for another stage in the pipeline
-//
 
 type L1TransactionFetcher interface {
 	InfoAndTxsByHash(ctx context.Context, hash common.Hash) (eth.BlockInfo, types.Transactions, error)
 }
 
-type DataSlice []eth.Data
+// CalldataSourceImpl is a fault tolerant approach to fetching data.
+// The constructor will never fail & it will instead re-attempt the fetcher
+// at a later point.
+// This API greatly simplifies some calling code.
+type CalldataSourceImpl struct {
+	// Internal state + data
+	open bool
+	data []eth.Data
+	// Required to re-attempt fetching
+	id      eth.BlockID
+	cfg     *rollup.Config // TODO: `DataFromEVMTransactions` should probably not take the full config
+	fetcher L1TransactionFetcher
+	log     log.Logger
+}
 
-func (ds *DataSlice) Next(ctx context.Context) (eth.Data, error) {
-	if len(*ds) == 0 {
+func NewCalldataSourceImpl(ctx context.Context, log log.Logger, cfg *rollup.Config, fetcher L1TransactionFetcher, block eth.BlockID) *CalldataSourceImpl {
+	_, txs, err := fetcher.InfoAndTxsByHash(ctx, block.Hash)
+	if err != nil {
```

Comment:
There should be a difference between temporary RPC errors and persistent not-found type of errors. Creating a calldata source with no data (if it doesn't error, but just returns nothing?) may result in accidental missed data, which could become a divergence. IMHO leaving the calldata as-is, and putting it in a separate PR or reducing changes is nice, since it's not quite the same as a derivation stage.